core/src/main/scala/cats/data/OrderedSet.scala

```diff
@@ -0,0 +1,231 @@
+package cats
+package data
+
+import cats.kernel._
+import kernel.compat.scalaVersionSpecific._
+import scala.annotation.tailrec
+import scala.collection.immutable.SortedSet
+
+/** A newtype over a `SortedSet` the construction of which is guaranteed to
+  * only use an [[Order]] instance, rather than `Ordering`. The intent here is
+  * that an `OrderedSet` will always use the coherent [[Order]] for a given
+  * type, implying for all `OrderedSet[A]` they will have the same
+  * ordering. If you bypass this restriction then your `OrderedSet` will be
+  * unlawful for many of the cats typeclasses.
+  *
+  * This differs from `SortedSet` in that any given instance of a `SortedSet`
+  * may be using any lawful, but not necessarily coherent, `Ordering`.
+  */
+private[data] object OrderedSetImpl extends OrderedSetInstances with Newtype {
+  private[data] def create[A](s: SortedSet[A]): Type[A] =
+    s.asInstanceOf[Type[A]]
+
+  private[data] def unwrap[A](s: Type[A]): SortedSet[A] =
+    s.asInstanceOf[SortedSet[A]]
+
+  def empty[A](implicit A: Order[A]): OrderedSet[A] =
+    create(SortedSet.empty(A.toOrdering))
+
+  def add[A](xs: OrderedSet[A], x: A): OrderedSet[A] =
+    create(unwrap(xs) + x)
+
+  def from[G[_], A](value: G[A])(implicit G: Foldable[G], A: Order[A]): OrderedSet[A] =
+    G.foldLeft(value, empty[A])(add)
+
+  def of[A: Order](x: A, xs: A*): OrderedSet[A] =
+    from((x +: xs).toList)
+
+  def apply[A: Order](x: A, xs: A*): OrderedSet[A] =
+    of(x, xs: _*)
+
+  def one[A: Order](x: A): OrderedSet[A] =
+    of(x)
+
+  implicit def catsSetOpsForOrderedSet[A](value: OrderedSet[A]): OrderedSetOps[A] =
+    new OrderedSetOps[A](value)
+}
+
+@suppressUnusedImportWarningForScalaVersionSpecific
+final class OrderedSetOps[A](override val set: OrderedSet[A]) extends SetOpsForOrderedSets[OrderedSet, OrderedSet, A] {
+
+  override def toSortedSet: SortedSet[A] =
+    OrderedSetImpl.unwrap(set)
+
+  def headOption: Option[A] =
+    toSortedSet.headOption
+
+  def lastOption: Option[A] =
+    toSortedSet.lastOption
+
+  def max: Option[A] =
+    lastOption
+
+  def min: Option[A] =
+    headOption
+
+  def minBy[B: Order](f: A => B): Option[A] =
+    if (toSortedSet.isEmpty) {
+      None
+    } else {
+      implicit val ordering: Ordering[B] = Order[B].toOrdering
+      Some(toSortedSet.minBy(f))
+    }
+
+  def maxBy[B: Order](f: A => B): Option[A] =
+    if (toSortedSet.isEmpty) {
+      None
+    } else {
+      implicit val ordering: Ordering[B] = Order[B].toOrdering
+      Some(toSortedSet.maxBy(f))
+    }
+
+  override def show(implicit A: Show[A]): String =
+    toSortedSet.iterator.map(A.show).mkString("OrderedSet(", ", ", ")")
+
+  override def ===(that: OrderedSet[A])(implicit A: Eq[A]): Boolean =
+    Eq[Int].eqv(length, that.length) && cats.kernel.instances.StaticMethods
```

Comment:
We don't need (and shouldn't use) `Eq[Int].eqv(a, b)` instead of `a == b`. It is just needlessly slower.

Reply:
If we proceed here, I'll update.